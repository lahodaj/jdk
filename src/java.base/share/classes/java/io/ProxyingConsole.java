/*
 * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.io;

import java.nio.charset.Charset;
import jdk.internal.io.JdkConsole;

/**
 * A proxy for the Console for internal use. Providers of a Console
 * can extend jdk.internal.io.JdkConsole class
 */

final class ProxyingConsole extends Console {
    private final JdkConsole delegate;

    ProxyingConsole(JdkConsole delegate) {
        this.delegate = delegate;
    }

    public PrintWriter writer() {
        return delegate.writer();
    }

    public Reader reader() {
        return delegate.reader();
    }

    public Console format(String fmt, Object ...args) {
        delegate.format(fmt, args);
        return this;
    }

    public Console printf(String format, Object ... args) {
        delegate.printf(format, args);
        return this;
    }

    public String readLine(String fmt, Object ... args) {
        return delegate.readLine(fmt, args);
    }

    public String readLine() {
        return delegate.readLine();
    }

    public char[] readPassword(String fmt, Object ... args) {
        return delegate.readPassword(fmt, args);
    }

    public char[] readPassword() {
        return delegate.readPassword();
    }

    public void flush() {
        delegate.flush();
    }

    public Charset charset() {
        return delegate.charset();
    }
}
