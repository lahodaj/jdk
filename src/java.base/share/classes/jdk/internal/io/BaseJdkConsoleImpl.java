/*
 * Copyright (c) 2022, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.io;

import java.io.IOError;
import java.io.IOException;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.io.Reader;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.Formatter;
import java.util.Locale;
import java.util.Objects;

import sun.nio.cs.StreamDecoder;
import sun.nio.cs.StreamEncoder;

/**
 * Base for JDK's JdkConsole implementations.
 */
public abstract class BaseJdkConsoleImpl implements JdkConsole {
    @Override
    public PrintWriter writer() {
        return pw;
    }

    @Override
    public Reader reader() {
        return reader;
    }

    @Override
    public JdkConsole println(Object obj) {
        pw.println(obj);
        // automatic flushing covers println
        return this;
    }

    @Override
    public JdkConsole print(Object obj) {
        pw.print(obj);
        pw.flush(); // automatic flushing does not cover print
        return this;
    }

    @Override
    public String readln(String prompt) {
        String line = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                pw.print(prompt);
                pw.flush(); // automatic flushing does not cover print
                try {
                    char[] ca = readline(false);
                    if (ca != null)
                        line = new String(ca);
                } catch (IOException x) {
                    throw new IOError(x);
                }
            }
        }
        return line;
    }

    @Override
    public String readln() {
        String line = null;
        synchronized(readLock) {
            try {
                char[] ca = readline(false);
                if (ca != null)
                    line = new String(ca);
            } catch (IOException x) {
                throw new IOError(x);
            }
        }
        return line;
    }

    @Override
    public JdkConsole format(Locale locale, String format, Object ... args) {
        formatter.format(locale, format, args).flush();
        return this;
    }

    @Override
    public String readLine(Locale locale, String format, Object ... args) {
        String line = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                if (!format.isEmpty())
                    pw.format(locale, format, args);
                try {
                    char[] ca = readline(false);
                    if (ca != null)
                        line = new String(ca);
                } catch (IOException x) {
                    throw new IOError(x);
                }
            }
        }
        return line;
    }

    @Override
    public String readLine() {
        return readLine(Locale.getDefault(Locale.Category.FORMAT), "");
    }

    @Override
    public char[] readPassword(Locale locale, String format, Object ... args) {
        char[] passwd = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                try {
                    if (!format.isEmpty())
                        pw.format(locale, format, args);
                    passwd = readline(true);
                } catch (IOException x) {
                    throw new IOError(x);
                }
                pw.println();
            }
        }
        return passwd;
    }

    @Override
    public char[] readPassword() {
        return readPassword(Locale.getDefault(Locale.Category.FORMAT), "");
    }

    @Override
    public void flush() {
        pw.flush();
    }

    @Override
    public Charset charset() {
        return charset;
    }

    protected Reader wrapReader(Reader baseReader) {
        return baseReader;
    }

    protected final Charset charset;
    protected final Object readLock;
    protected final Object writeLock;
    protected final Reader reader;
    protected final Writer out;
    protected final PrintWriter pw;
    protected final Formatter formatter;

    protected abstract char[] readline(boolean password) throws IOException;

    @SuppressWarnings("this-escape")
    public BaseJdkConsoleImpl(Charset charset) {
        Objects.requireNonNull(charset);
        this.charset = charset;
        readLock = new Object();
        writeLock = new Object();
        out = StreamEncoder.forOutputStreamWriter(
                new FileOutputStream(FileDescriptor.out),
                writeLock,
                charset);
        pw = new PrintWriter(out, true) {
            public void close() {
            }
        };
        formatter = new Formatter(out);
        StreamDecoder plainReader = StreamDecoder.forInputStreamReader(
                new FileInputStream(FileDescriptor.in),
                readLock,
                charset);
        reader = wrapReader(plainReader);
    }
}
