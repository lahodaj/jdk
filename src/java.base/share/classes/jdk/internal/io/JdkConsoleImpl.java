/*
 * Copyright (c) 2022, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.io;

import java.io.IOError;
import java.io.IOException;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.io.Reader;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.Formatter;
import java.util.Locale;
import java.util.Objects;

import jdk.internal.access.SharedSecrets;
import sun.nio.cs.StreamDecoder;
import sun.nio.cs.StreamEncoder;

/**
 * JdkConsole implementation based on the platform's TTY.
 */
public final class JdkConsoleImpl implements JdkConsole {
    @Override
    public PrintWriter writer() {
        return pw;
    }

    @Override
    public Reader reader() {
        return reader;
    }

    @Override
    public JdkConsole println(Object obj) {
        pw.println(obj);
        // automatic flushing covers println
        return this;
    }

    @Override
    public JdkConsole print(Object obj) {
        pw.print(obj);
        pw.flush(); // automatic flushing does not cover print
        return this;
    }

    @Override
    public String readln(String prompt) {
        String line = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                pw.print(prompt);
                pw.flush(); // automatic flushing does not cover print
                try {
                    char[] ca = readline(false);
                    if (ca != null)
                        line = new String(ca);
                } catch (IOException x) {
                    throw new IOError(x);
                }
            }
        }
        return line;
    }

    @Override
    public String readln() {
        String line = null;
        synchronized(readLock) {
            try {
                char[] ca = readline(false);
                if (ca != null)
                    line = new String(ca);
            } catch (IOException x) {
                throw new IOError(x);
            }
        }
        return line;
    }

    @Override
    public JdkConsole format(Locale locale, String format, Object ... args) {
        formatter.format(locale, format, args).flush();
        return this;
    }

    @Override
    public String readLine(Locale locale, String format, Object ... args) {
        String line = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                if (!format.isEmpty())
                    pw.format(locale, format, args);
                try {
                    char[] ca = readline(false);
                    if (ca != null)
                        line = new String(ca);
                } catch (IOException x) {
                    throw new IOError(x);
                }
            }
        }
        return line;
    }

    @Override
    public String readLine() {
        return readLine(Locale.getDefault(Locale.Category.FORMAT), "");
    }

    @Override
    public char[] readPassword(Locale locale, String format, Object ... args) {
        char[] passwd = null;
        synchronized (writeLock) {
            synchronized(readLock) {
                installShutdownHook();
                try {
                    synchronized(restoreEchoLock) {
                        restoreEcho = echo(false);
                    }
                } catch (IOException x) {
                    throw new IOError(x);
                }
                IOError ioe = null;
                try {
                    if (!format.isEmpty())
                        pw.format(locale, format, args);
                    passwd = readline(true);
                } catch (IOException x) {
                    ioe = new IOError(x);
                } finally {
                    try {
                        synchronized(restoreEchoLock) {
                            if (restoreEcho) {
                                restoreEcho = echo(true);
                            }
                        }
                    } catch (IOException x) {
                        if (ioe == null)
                            ioe = new IOError(x);
                        else
                            ioe.addSuppressed(x);
                    }
                    if (ioe != null) {
                        //TODO:
//                        Arrays.fill(passwd, ' ');
//                        try {
//                            if (reader instanceof LineReader lr) {
//                                lr.zeroOut();
//                            }
//                        } catch (IOException _) {
//                            // ignore
//                        }
                        throw ioe;
                    }
                }
                pw.println();
            }
        }
        return passwd;
    }

    private void installShutdownHook() {
        if (shutdownHookInstalled)
            return;
        try {
            // Add a shutdown hook to restore console's echo state should
            // it be necessary.
            SharedSecrets.getJavaLangAccess()
                    .registerShutdownHook(0 /* shutdown hook invocation order */,
                            false /* only register if shutdown is not in progress */,
                            new Runnable() {
                                public void run() {
                                    try {
                                        synchronized(restoreEchoLock) {
                                            if (restoreEcho) {
                                                echo(true);
                                            }
                                        }
                                    } catch (IOException _) { }
                                }
                            });
        } catch (IllegalStateException _) {
            // shutdown is already in progress and readPassword is first used
            // by a shutdown hook
        }
        shutdownHookInstalled = true;
    }

    @Override
    public char[] readPassword() {
        return readPassword(Locale.getDefault(Locale.Category.FORMAT), "");
    }

    @Override
    public void flush() {
        pw.flush();
    }

    @Override
    public Charset charset() {
        return charset;
    }

    private final Charset charset;
    private final Object readLock;
    private final Object writeLock;
    // Must not block while holding this. It is used in the shutdown hook.
    private final Object restoreEchoLock;
    private final Reader reader;
    private final Writer out;
    private final NativeConsoleReader nativeConsoleReader;
    private final PrintWriter pw;
    private final Formatter formatter;
    private boolean restoreEcho;
    private boolean shutdownHookInstalled;

    private char[] readline(boolean zeroOut) throws IOException {
        return nativeConsoleReader.readline(reader, out, zeroOut);
    }

    /*
     * Sets the console echo status to {@code on} and returns the previous
     * console on/off status.
     * @param on    the echo status to set to. {@code true} for echo on and
     *              {@code false} for echo off
     * @return true if the previous console echo status is on
     */
    private static native boolean echo(boolean on) throws IOException;

    public JdkConsoleImpl(Charset charset) {
        Objects.requireNonNull(charset);
        this.charset = charset;
        readLock = new Object();
        writeLock = new Object();
        restoreEchoLock = new Object();
        out = StreamEncoder.forOutputStreamWriter(
                new FileOutputStream(FileDescriptor.out),
                writeLock,
                charset);
        pw = new PrintWriter(out, true) {
            public void close() {
            }
        };
        formatter = new Formatter(out);
        reader = /*new LineReader(*/StreamDecoder.forInputStreamReader(
                new FileInputStream(FileDescriptor.in),
                readLock,
                charset);//);
        nativeConsoleReader = NativeConsoleReaderImpl.create(readLock);
    }

}
