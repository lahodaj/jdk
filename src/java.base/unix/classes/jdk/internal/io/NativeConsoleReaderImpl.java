/*
 * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package jdk.internal.io;

import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import jdk.internal.io.SimpleConsoleReader.Size;

public class NativeConsoleReaderImpl implements NativeConsoleReader {

    public static NativeConsoleReader create(Object readLock) {
        if (CLibrary.isTty(0)) {
            return new NativeConsoleReaderImpl();
        } else {
            return new BaseNativeConsoleReader(readLock);
        }
    }

    //TODO: read lock
    public char[] readline(Reader reader, Writer out, boolean zeroOut) throws IOException {
        Attributes originalAttributes = CLibrary.getAttributes(0);
        Attributes rawAttributes = new Attributes(originalAttributes);
        rawAttributes.setInputFlag(Attributes.InputFlag.BRKINT, false);
        rawAttributes.setInputFlag(Attributes.InputFlag.IGNPAR, false);
        rawAttributes.setInputFlag(Attributes.InputFlag.ICRNL, false);
        rawAttributes.setInputFlag(Attributes.InputFlag.IXON, false);
        rawAttributes.setInputFlag(Attributes.InputFlag.IXOFF, true);
        rawAttributes.setInputFlag(Attributes.InputFlag.IMAXBEL, false);
        rawAttributes.setLocalFlag(Attributes.LocalFlag.ICANON, false);
        rawAttributes.setLocalFlag(Attributes.LocalFlag.ECHO, false);
        Thread restoreConsole = new Thread(() -> {
            CLibrary.setAttributes(0, originalAttributes);
        });
        try {
            Runtime.getRuntime().addShutdownHook(restoreConsole);
            CLibrary.setAttributes(0, rawAttributes);
            Size size = CLibrary.getTerminalSize(0);
            out.append("\033[6n").flush(); //ask the terminal to provide cursor location
            return SimpleConsoleReader.doRead(reader, out, -1, () -> size.width());
        } finally {
            restoreConsole.run();
            Runtime.getRuntime().removeShutdownHook(restoreConsole);
        }
    }

}
