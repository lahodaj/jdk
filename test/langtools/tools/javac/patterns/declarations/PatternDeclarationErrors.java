import java.io.IOException;

/*
 * @test /nodynamiccopyright/
 * @summary Verify error related to annotations and patterns
 * @enablePreview
 * @compile/fail/ref=PatternDeclarationErrors.out -XDrawDiagnostics -XDdev --should-stop=at=FLOW PatternDeclarationErrors.java
 */
public class PatternDeclarationErrors {
    public class D {
        public pattern D(Object v1, Float out) {
            match D(10.0f); // not matching signature
        }

        public pattern D(Float out, Integer v1) {
            match D2(10.0f, 2); // no matching name of pattern declaration
        }

        public pattern D(Float out, Double v1) {
            match D(10.0f, "2"); // inconvertible types
        }

        public void D3(Float out, Integer v1) {
            match D3(10.0f, 2); // match in regular method
        }

        public pattern D(Float out, Character v1) {
            match D(10.0f, '2');
            System.out.println("unreachable"); // unreachable
        }
    }

    public class ExitErrors {
        public pattern ExitErrors(int out) {
           break; // no break
        }

        public pattern ExitErrors(float out) {
            return Integer.valueOf(42); // no return
        }

        public pattern ExitErrors(char out) {
            yield Integer.valueOf(42); // no yield
        }
    }

    public class ExceptionErrors {
        public pattern ExceptionErrors(int out) {
            throw new Error(); // no throws in patter declaration body
        }

        public pattern ExceptionErrors(float out) throws IOException { // no throws in pattern signature
            throw new IOException(); // no throws in patter declaration body
        }
    }

    public record R(int i) {
        public pattern R(int i) { match R(42); }
    }

    public class R1 {
        public pattern R1(int i) { match R1(42); }
        public pattern R1(int i) { match R1(42); }
    }

    public class R2 {
        pattern R2(int i) { match R2(42); }
        pattern R2(int i) { match R2(42); }
    }


    public class Point {
        final int x;
        final int y;

        protected Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        protected pattern Point(int x, int y) {
            match Point(this.x, this.y);
        }
    }

    public class GreatPoint extends Point {
        final int magnitude;

        public GreatPoint(int x, int y, int magnitude) {
            super(x, y);
            if (magnitude < 0) throw new IllegalArgumentException();
            this.magnitude = magnitude;
        }

        public pattern GreatPoint(int x, int y, int magnitude) {
            if (this instanceof Point(var X, var Y)) {
                x = X;
                y = Y;
            }
            match GreatPoint(x, y, this.magnitude);
        }
    }
}
