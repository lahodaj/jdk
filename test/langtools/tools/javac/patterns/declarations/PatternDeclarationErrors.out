PatternDeclarationErrors.java:63:24: compiler.err.already.defined: kindname.pattern.declaration, R1(), kindname.class, PatternDeclarationErrors.R1
PatternDeclarationErrors.java:68:17: compiler.err.already.defined: kindname.pattern.declaration, R2(), kindname.class, PatternDeclarationErrors.R2
PatternDeclarationErrors.java:12:13: compiler.err.match.not.matching.pattern.declaration.signature
PatternDeclarationErrors.java:16:13: compiler.err.match.pattern.name.wrong
PatternDeclarationErrors.java:20:28: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.String, java.lang.Double)
PatternDeclarationErrors.java:24:13: compiler.err.match.outside.pattern.declaration
PatternDeclarationErrors.java:35:12: compiler.err.break.outside.switch.loop
PatternDeclarationErrors.java:39:13: compiler.err.ret.outside.meth
PatternDeclarationErrors.java:43:13: compiler.err.no.switch.expression
PatternDeclarationErrors.java:49:13: compiler.err.pattern.declaration.no.throws
PatternDeclarationErrors.java:52:24: compiler.err.pattern.declaration.cant.throw.exception
PatternDeclarationErrors.java:53:13: compiler.err.pattern.declaration.no.throws
PatternDeclarationErrors.java:58:24: compiler.err.invalid.canonical.deconstructor.in.record: R
PatternDeclarationErrors.java:97:17: compiler.err.cant.assign.val.to.var: final, x
PatternDeclarationErrors.java:98:17: compiler.err.cant.assign.val.to.var: final, y
PatternDeclarationErrors.java:29:13: compiler.err.unreachable.stmt
- compiler.note.preview.filename: PatternDeclarationErrors.java, DEFAULT
- compiler.note.preview.recompile
16 errors